### PR TITLE
CCM-2152: Remove aws headers from responses

### PR DESCRIPTION
**Pull request by TheWorldWideWebb**
## Summary
This work aims to remove amazon headers from the response, Ive taken a regex approach to pattern match on x-amz. I have extended the testing assertion library to assert on 3 different types on headers which also covers the majority of edge cases e.g. casing and amzn type headers.

## Reviews Required
* [x] Dev
* [x] Test
* [ ] Tech Author
* [ ] Product Owner

## Checklist
* [ ] Brief description of work completed, and any technical decisions made as part of the PR
* [ ] PR link added as a comment to the relevant JIRA ticket
* [ ] Branch deployed to a dynamic env - link to deployment pipeline
* [ ] PR link shared on Slack and/or Teams
* [ ] 1 reviews received

### CODE REVIEW

File: tests/sandbox/headers/test_x_amz_removal.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+
+@pytest.mark.sandboxtest
+@pytest.mark.parametrize("method", TEST_METHODS)
+@pytest.mark.parametrize("endpoints", VALID_ENDPOINTS)
+def test_request_with_x_amz_is_removed(nhsd_apim_proxy_url, endpoints, method):
+
+    resp = getattr(requests, method)(f"{nhsd_apim_proxy_url}/{endpoints}", headers={
+        "Accept": "*/*",
+        "Origin": "https://my.website"
```
Comment (Kennedy1265) [inline tests/sandbox/headers/test_x_amz_removal.py:15]: Looks like you've created an ORIGIN constant that could be used here, there's also a couple other instances in the tests that directly set https://my.website that could be replaced as well under cors tests if you're going this route

Comment (TheWorldWideWebb) [reply tests/sandbox/headers/test_x_amz_removal.py]: Made consistent across tests

**Review by Kennedy1265 — APPROVED** [review]
Testing looks good

File: proxies/shared/resources/jsc/SetResponseDefaults.js [hunk 2]
```diff
@@ -18,6 +18,17 @@ context.setVariable("error.header.Cache-Control", "no-cache, no-store, must-reva
 context.setVariable("response.header.X-Content-Type-Options", "nosniff");
 context.setVariable("error.header.X-Content-Type-Options", "nosniff");
 
+// remove aws headers
+
```
Comment (PhilipSkinner) [inline proxies/shared/resources/jsc/SetResponseDefaults.js:22]: remove the extra `\n` please

File: tests/lib/constants/constants.py [hunk 3]
```diff
@@ -26,7 +26,9 @@
 
 TOKENS = [None, "Bearer xyzcba", "Bearer", "junk"]
 METHODS = ["get", "post", "put", "patch", "delete", "head", "options"]
+TEST_METHODS = ["get", "post", "put", "delete"]
```
Comment (PhilipSkinner) [inline tests/lib/constants/constants.py:29]: We should be using the existing set of `METHODS` for these tests I believe.

File: tests/production/headers/test_cors.py [hunk 4]
```diff
@@ -1,8 +1,7 @@
 import requests
 import pytest
 from lib import Assertions, Authentication
-from lib.constants.constants import PROD_URL, VALID_ENDPOINTS
-
+from lib.constants.constants import PROD_URL, VALID_ENDPOINTS, ORIGIN
 
 METHODS = ["GET", "POST", "PUT", "PATCH", "DELETE"]
 TEST_METHODS = ["get", "post", "put", "delete"]
```
Comment (PhilipSkinner) [inline tests/production/headers/test_cors.py:7]: hmmm, why do we have these two lists of methods?

**Review by PhilipSkinner — CHANGES_REQUESTED** [review]
Please see comments, once resolved will give this a tick!

File: tests/lib/constants/constants.py (hunk 3 above)
Comment (TheWorldWideWebb) [reply tests/lib/constants/constants.py:29]: I would only be inclined to agree if someone could give a reason as to why they are separate currently, or if we can agree as a consensus that we should migrate them to these constants.

File: tests/production/headers/test_cors.py (hunk 4 above)
Comment (TheWorldWideWebb) [reply tests/production/headers/test_cors.py:7]: As above, there existence is unclear

File: proxies/shared/resources/jsc/SetResponseDefaults.js [hunk 5]
```diff
@@ -18,6 +18,17 @@ context.setVariable("error.header.Cache-Control", "no-cache, no-store, must-reva
...
+const headerNames = (context.getVariable("response.headers.names") + "").slice(1, -1).split(', ');
```
[the remark this answers is not in the document]
Comment (TheWorldWideWebb) [reply proxies/shared/resources/jsc/SetResponseDefaults.js:23]: Is that driven from the backend lambdas populating them?

[the remark this answers is not in the document]
Comment (TheWorldWideWebb) [reply proxies/shared/resources/jsc/SetResponseDefaults.js:23]: I did find this which would suggest we dont need to interact with the error headers at all but let me know if this error object is not the same thing

https://docs.apigee.com/api-platform/reference/variables-reference#error
https://docs.apigee.com/api-platform/fundamentals/fault-handling

File: tests/lib/constants/constants.py (hunk 3 above)
Comment (TheWorldWideWebb) [reply tests/lib/constants/constants.py:29]: Ive tried running the tests with the additional methods but was met with failures for the options method

File: proxies/shared/resources/jsc/SetResponseDefaults.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (PhilipSkinner) [reply proxies/shared/resources/jsc/SetResponseDefaults.js:23]: Yeah I think you're right, I'm mistaken about this - your code looks good and we can see it working with a backend thrown exception as it correctly removes the AWS headers from it.

**Review by PhilipSkinner — APPROVED** [review]
Approved!
